### ISSUE TITLE
UI keybinding for specific audio player

### ISSUE THREAD

**Issue by spurkopf** [zynthian/zynthian-issue-tracking#906]
in short:
Please add an additonal UI keybinding like Start_Audio_Play for the new audio player when used as special chain inside a snapshot.

long story:
I am currently trying to create a live setup with the zynthian. The midikeyboard is limited in its functions, so I use an additional numeric keypad on the zynthian. I use one snapshot per song, and one audio player in each. It would be great if play/pause/stop could be triggered with the numeric keypad.

**Comment by riban-bw** [comment]
This is already possible with keybinding which is configured in webconf INTERFACE->UI Keybinding. You can bind a key to toggle playback. Note that this toggles the currently showing (editor screen) audio player. If none are shown then the global audio player (that usually replays the last capture recording) is controlled.

**Comment by spurkopf** [comment]
Yes, the UI keybinding for the global audio player is known, but it is not a solution to the problem.

**Comment by riban-bw** [comment]
Your request was for play/pause/stop to be triggered by a numeric keypad. I did exactly this today with key binding. I don't understand what is not working.

**Comment by spurkopf** [comment]
Thank you for thinking about this again.  Perhaps this function of the audio player is not even known? 
That´s what I do:
I place one audio player in each snapshot - from the menu Add Special Chain / Audio File Player, 
I select a Midi channel and can then select a **different** .wav audio file. 
This selection is also saved in the snapshot. 
When I load that snapshot, I have the corresponding .wav file as a backing track together with the matching sounds for the midi keyboard. 

With the global audio player I could just play the same song over and over again, nobody would book me for that ;-)

**Comment by riban-bw** [comment]
- Add a chain with an Audio Player
- Load an audio file into the chain's audio player
- Open the chain's control view to show the waveform
- Use keybinding to driver the chain's audio player

Maybe you want a different behaviour but this is exactly how to control a chain's audio player from keybinding.

**Comment by spurkopf** [comment]
Sorry, I cannot confirm this functionality. The START_AUDIO_PLAY keybinding only activates the global audio player, even if the "Waveform view" is selected. Regardless, even if it worked this way, it would be impractical. I select the different snapshots with numeric keypad, after loading is done - Zynthian goes to the mixer view.

**Comment by riban-bw** [comment]
I'm not near a Zynthian at the moment but the binding your want is something like TOGGLE_PLAY.

**Comment by spurkopf** [comment]
I have just tried it, TOGGLE_AUDIO_PLAY also always plays the same audio file, regardless of which snapshot is loaded with Chain-Audio Player and regardless of whether the "waveform view" is displayed.

**Comment by spurkopf** [comment]
In the meantime I think there is something wrong with the audio player integration in general. But I think you are working on integrating audio tracks into the sequencer anyway. It is probably better to wait for a well thought out solution. (I'll ask for a UI binding for play/stop/toggle for that ;-) ) 
In the meantime I'm making do by sacrificing the drum pads on my live keyboard for the play/return to start function.

**Comment by riban-bw** [comment]
I didn't say, "TOGGLE_AUDIO_PLAY", I said "TOGGLE_PLAY". These are different CUIA commands. "TOGGLE_PLAY" will toggle the currently showing player which might be an audio player, MIDI player, etc. This CUIA works perfectly for me. I have keybinding for '0' key on my numeric pad to "TOGGLE_PLAY"  and the button will start and pause playback.

![image](https://github.com/zynthian/zynthian-issue-tracking/assets/3158323/7dce4c75-7cbc-4bf2-8416-654172088834)

You can also map "STOP" to a button to stop and reset to start of audio.

**Comment by spurkopf** [comment]
Yes, TOGGLE_PLAY works in this way, but it doesn't help.

**Comment by riban-bw** [comment]
I'm sorry but I can't help you if you don't tell me why this doesn't help. It seems to meet the requiremment of being able to start / stop / pause playback of individual audio players in seperate chains.

**Comment by spurkopf** [comment]
Sorry, I thought this was already explained in the answer above. The Zynthian is in the mixer view after loading a snapshot. If I had time on stage to run to the Zynthian after every song and then switch the menu pages there directly, it would be no problem to turn the back encoder slightly to start the audio player.
The Zynthian must be controlled completely wirelessly on stage.

**Comment by riban-bw** [comment]
I have changed the title of the ticket. I believe the requirement is to be able to toggle the play state of a specific audio player. The player may be in a chain in which case the chain's ID may be a possible identifier for the command. Each player will also have a UID but this is dynamically created so cannot be used as a persistent identifier. The player will have a _processor id_ which can be saved in the snapshot so that could possibly be used. Also the player will be in a list of players but that list may be created in different ways depending on how snapshots are recalled or manual adding and removing players happens.

So as I mentioned early in this ticket, it may be possible to add a parameter to the toggle_play cuia message to target the player. I will reopen the ticket whilst we reconsider if that is a viable and desirbale option. (What was said before about ensuring this scales still applies.)

**Comment by riban-bw** [comment]
I have a suggestion that may be suboptimal but may allow the required functionality, even as a temporary work-around:

- Bind a keypad button to CUIA "CHAIN_CONTROL" with a single parameter of the index of the chain containing the audio player, e.g. "1" for the first chain.
- Bind a keypad button to CUIA "TOGGLE_PLAY"
- After loading snapshot, press the first button to show the player control view
- Press the second button to toggle play

**Comment by spurkopf** [comment]
Of course I don't know what will happen with the Zynthian software, but if the midi channel is retained in the chain-audio-player, it might also be a solution to control the corresponding player via the midi channel. This works well for me for the program changes via the global midi channel to load snapshots.
Your workaround with CHAIN_CONTROL + TOGGLE_PLAY would probably work, but at the moment I use my workaround with the drum pads and Midi Learn to control the player, which also works satisfactorily.
